### PR TITLE
feat: check if the service is used by route when deleting

### PR DESCRIPTION
**Pull request by starsz**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [ ] Bugfix
- [x] New feature provided
- [ ] Improve performance
- [ ] Backport patches

### New feature or improvement
- Describe the details and related test reports.

Now, when deleting a service, we don't check whether there is a route using it.
This may lead to the DP error if the user deletes the service by mistake.

So, we should check if the service is using by route when deleting.
If the service is using, we return 400 to guarantee the DP works well.

___
### Please add the corresponding test cases if necessary.

Added.
___

### CODE REVIEW

**Review by nic-chen — APPROVED** [review]

File: api/internal/handler/service/service.go [hunk 1]
```diff
@@ -218,8 +220,34 @@ type BatchDelete struct {
...
+		Predicate: func(obj interface{}) bool {
+			route := obj.(*entity.Route)
+			if _, exist := mp[utils.InterfaceToString(route.ServiceID)]; exist {
+				return true
+			}
+
+			return false
+		},
+		PageSize:   0,
+		PageNumber: 0,
```
Comment (gxthrj) [inline api/internal/handler/service/service.go:239]: why both `pagesize` and `pagenumber` are 0 ?

Comment (starsz) [reply api/internal/handler/service/service.go:239]: We need to list all routes.

File: api/test/e2enew/service/service_test.go [hunk 2]
```diff
@@ -462,3 +463,83 @@ var _ = ginkgo.Describe("service update use patch method", func() {
...
+					"port":   1980,
+					"weight": 1,
+				},
+			},
+		},
+	}
+	_createServiceBody, err := json.Marshal(createServiceBody)
+	assert.Nil(t, err)
+
+	table.DescribeTable("test service delete",
```
Comment (moonming) [inline api/test/e2enew/service/service_test.go:485]: is there any test for DP？

**Review by moonming — CHANGES_REQUESTED** [review]

File: api/test/e2enew/service/service_test.go [hunk 3]
```diff
@@ -82,9 +83,9 @@ var _ = ginkgo.Describe("create service without plugin", func() {
 			Method: http.MethodPut,
 			Path:   "/apisix/admin/routes/r1",
 			Body: `{
-				"uri": "/server_port",
-				"service_id": "s1"
-			}`,
+ 				"uri": "/server_port",
```
Comment (tokers) [inline api/test/e2enew/service/service_test.go:86]: Seems that you added some empty spaces?

File: api/test/e2enew/service/service_test.go (hunk 2 above)
Comment (starsz) [reply api/test/e2enew/service/service_test.go:485]: Yes.The DP test are here, you can see:

https://github.com/apache/apisix-dashboard/blob/01c952803d256c53ac411ca52d1c90a2263879fc/api/test/e2e/route_service_upstream_test.go#L93

File: api/test/e2enew/service/service_test.go (hunk 3 above)
Comment (starsz) [reply api/test/e2enew/service/service_test.go:86]: I can't found it out.

File: api/test/e2enew/service/service_test.go (hunk 2 above)
Comment (moonming) [reply api/test/e2enew/service/service_test.go:485]: but not in this file

Comment (starsz) [reply api/test/e2enew/service/service_test.go:485]: And here, there are in this file: 

https://github.com/apache/apisix-dashboard/blob/01c952803d256c53ac411ca52d1c90a2263879fc/api/test/e2enew/service/service_test.go#L93

https://github.com/apache/apisix-dashboard/blob/01c952803d256c53ac411ca52d1c90a2263879fc/api/test/e2enew/service/service_test.go#L193

https://github.com/apache/apisix-dashboard/blob/01c952803d256c53ac411ca52d1c90a2263879fc/api/test/e2enew/service/service_test.go#L310

File: api/test/e2enew/service/service_test.go (hunk 3 above)
Comment (starsz) [reply api/test/e2enew/service/service_test.go:86]: Let me check the difference.

File: api/test/e2enew/service/service_test.go (hunk 2 above)
Comment (moonming) [reply api/test/e2enew/service/service_test.go:485]: not test DP for this new test case

File: api/test/e2enew/service/service_test.go [hunk 4]
```diff
@@ -462,3 +463,83 @@ var _ = ginkgo.Describe("service update use patch method", func() {
...
+			ExpectBody:   "route: route1 is using this service",
+		}),
+		table.Entry("delete route first", base.HttpTestCase{
+			Desc:         "delete route first",
+			Object:       base.ManagerApiExpect(),
+			Method:       http.MethodDelete,
+			Path:         "/apisix/admin/routes/r1",
+			Headers:      map[string]string{"Authorization": base.GetToken()},
+			ExpectStatus: http.StatusOK,
+		}),
```
Comment (Jaycean) [inline api/test/e2enew/service/service_test.go:536]: IMHO, we should add test cases that verify the data is deleted after deleting route data and services data

Comment (moonming) [reply api/test/e2enew/service/service_test.go:536]: big +1

File: api/test/e2enew/service/service_test.go (hunk 2 above)
Comment (starsz) [reply api/test/e2enew/service/service_test.go:485]: IMO, different test cases test different features.
The former case tests the feature of creating a service and test it in DP.
This test case is testing for the relation of service and upstream (we can't delete service when the service is used by routes.

If this case tests the DP, I think the former test doesn't have mean to exist.

File: api/test/e2enew/service/service_test.go (hunk 4 above)
Comment (starsz) [reply api/test/e2enew/service/service_test.go:536]: Good capture. Let me add it.

File: api/test/e2enew/service/service_test.go (hunk 3 above)
Comment (starsz) [reply api/test/e2enew/service/service_test.go:86]: Change the space to tab.

**Comment by starsz** [comment]
@LiteSun Please have a look at Frontend CI.

**Comment by LiteSun** [comment]
> @LiteSun Please have a look at Frontend CI.

ok

**Comment by nic-chen** [comment]
@starsz 
I think you should sync codes from branch master to see if there is conflict in e2e test cases.

**Comment by starsz** [comment]
> @starsz
> I think you should sync codes from branch master to see if there is conflict in e2e test cases.

I think @LiteSun had done this before.

File: api/test/e2enew/service/service_test.go (hunk 2 above)
Comment (starsz) [reply api/test/e2enew/service/service_test.go:541]: Add test DP for this case.

**Review by tokers — APPROVED** [review]

**Review by nic-chen — APPROVED** [review]

**Review by gxthrj — APPROVED** [review]

**Review by moonming — APPROVED** [review]

File: api/internal/handler/service/service.go (hunk 1 above)
Comment (juzhiyuan) [reply api/internal/handler/service/service.go:253]: `0` means all?

Comment (juzhiyuan) [reply api/internal/handler/service/service.go:253]: In restful APIs, when we omit `size` or `number`, it means to fetch all resources. If we pass `0`, there should return an empty array IMO.

cc @imjoey @tokers to confirm

Comment (juzhiyuan) [reply api/internal/handler/service/service.go:253]: `offset = abs(page - 1) * pageSize`

**Review by juzhiyuan — COMMENTED** [review]
Here has one issue:

```
ret, err := h.routeStore.List(c.Context(), store.ListInput{
		Predicate: func(obj interface{}) bool {
			route := obj.(*entity.Route)
			if _, exist := mp[utils.InterfaceToString(route.ServiceID)]; exist {
				return true
			}

			return false
		},
		PageSize:   0,
		PageNumber: 0,
```

File: api/internal/handler/service/service.go (hunk 1 above)
Comment (starsz) [reply api/internal/handler/service/service.go:253]: You can see here:

https://github.com/apache/apisix-dashboard/blob/38cf0ba32264d5a7bb4f22bd40ce037e2f66a939/api/internal/core/store/store.go#L210-L225

**Comment by juzhiyuan** [comment]
cc @guoqqqi to check this failed CI

**Pull request merged**